core/string/string_name.h

```diff
@@ -165,24 +165,49 @@ class StringName {
 	static StringName search(const String &p_name);
 
 	struct AlphCompare {
-		_FORCE_INLINE_ bool operator()(const StringName &l, const StringName &r) const {
+		template <typename LT, typename RT>
+		_FORCE_INLINE_ bool operator()(const LT &l, const RT &r) const {
+			return compare(l, r);
+		}
+		_FORCE_INLINE_ static bool compare(const StringName &l, const StringName &r) {
```

Comment:
You didn't write all this, but while you're rewriting anyway, I suggest simplifying this logic as such:

```c++
			const char *l_cname = l._data ? l._data->cname : nullptr;
 			const char *r_cname = r._data ? r._data->cname : nullptr;

 			if (!r_cname) {
                return false;
            }
            if (!l_cname) {
                return true;
            }
            return str_compare(l_cname, r_cname) < 0;
```

And analogously for the other combinations. This should result in slightly faster, easier to read code :)